{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "432a7565",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import gzip\n",
    "import os\n",
    "import scipy.io\n",
    "import os.path\n",
    "from pathlib import Path\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "3150a482",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to your CSV file\n",
    "csv_file_path = 'tissue_positions.csv'\n",
    "\n",
    "# Making an empty list\n",
    "barcode = []\n",
    "\n",
    "# Open the CSV file in read mode\n",
    "with open(csv_file_path, 'r') as csv_file:\n",
    "    # Create a CSV reader object\n",
    "    csv_reader = csv.reader(csv_file)\n",
    "    \n",
    "    # Skipping column header\n",
    "    next(csv_reader)\n",
    "    \n",
    "    # Iterate over each row in the CSV file\n",
    "    for row in csv_reader:\n",
    "        # Each row is a list of values\n",
    "        barcode.append(row)\n",
    "\n",
    "barcode_coord = list(map(lambda x: [x[0], x[1], x[2], x[3]], barcode))\n",
    "barcode_wcell = list(filter(lambda x: x[1] == '1', barcode_coord))\n",
    "\n",
    "# Checking\n",
    "#barcode_wcell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "72052b30",
   "metadata": {},
   "outputs": [],
   "source": [
    "barcode_dict = {}\n",
    "for dots in barcode_wcell:\n",
    "    if dots[0] != barcode_dict:\n",
    "        barcode_dict[dots[0]] = dots[2] + ', ' + dots[3]\n",
    "\n",
    "# Checking\n",
    "#barcode_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "dc5dd078",
   "metadata": {},
   "outputs": [],
   "source": [
    "## This block will take a while to run ##\n",
    "cd = Path(os.path.abspath(\"\"))\n",
    "\n",
    "# define MEX directory\n",
    "matrix_dir = os.path.join(cd, \"CytAssist_FFPE_Protein_Expression_Human_Glioblastoma_filtered_feature_bc_matrix\", \"filtered_feature_bc_matrix\")\n",
    "# read in MEX format matrix as table\n",
    "mat_filtered = scipy.io.mmread(os.path.join(matrix_dir,\"matrix.mtx.gz\"))\n",
    "\n",
    "# list of transcript ids, e.g. 'ENSG00000187634'\n",
    "features_path = os.path.join(matrix_dir, \"features.tsv.gz\")\n",
    "feature_ids = [row[0]  for row  in csv.reader(gzip.open(features_path, mode=\"rt\"), delimiter=\"\\t\")]\n",
    "\n",
    "# list of gene names, e.g. 'SAMD11'\n",
    "gene_names = [row[1]  for row  in csv.reader(gzip.open(features_path, mode=\"rt\"), delimiter=\"\\t\")]\n",
    "\n",
    "# list of feature_types, e.g. 'Gene Expression'\n",
    "feature_types = [row[2]  for row  in csv.reader(gzip.open(features_path, mode=\"rt\"), delimiter=\"\\t\")]\n",
    "\n",
    "# list of barcodes, e.g. 'AAACATACAAAACG-1'\n",
    "barcodes_path = os.path.join(matrix_dir, \"barcodes.tsv.gz\")\n",
    "barcodes = [row[0]  for row  in csv.reader(gzip.open(barcodes_path, mode=\"rt\"), delimiter=\"\\t\")]\n",
    "barcodes = [barcode_dict[i] for i in barcodes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "7dff3931",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature id</th>\n",
       "      <th>gene name</th>\n",
       "      <th>54, 32</th>\n",
       "      <th>98, 158</th>\n",
       "      <th>31, 39</th>\n",
       "      <th>53, 139</th>\n",
       "      <th>34, 176</th>\n",
       "      <th>56, 142</th>\n",
       "      <th>89, 127</th>\n",
       "      <th>104, 118</th>\n",
       "      <th>...</th>\n",
       "      <th>65, 153</th>\n",
       "      <th>49, 133</th>\n",
       "      <th>73, 69</th>\n",
       "      <th>84, 102</th>\n",
       "      <th>71, 155</th>\n",
       "      <th>40, 78</th>\n",
       "      <th>24, 108</th>\n",
       "      <th>37, 127</th>\n",
       "      <th>83, 87</th>\n",
       "      <th>86, 106</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000187961</td>\n",
       "      <td>KLHL17</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000187583</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000187642</td>\n",
       "      <td>PERM1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18115</th>\n",
       "      <td>CD4</td>\n",
       "      <td>CD4</td>\n",
       "      <td>68679</td>\n",
       "      <td>33584</td>\n",
       "      <td>39114</td>\n",
       "      <td>54248</td>\n",
       "      <td>48609</td>\n",
       "      <td>56826</td>\n",
       "      <td>61884</td>\n",
       "      <td>31830</td>\n",
       "      <td>...</td>\n",
       "      <td>43931</td>\n",
       "      <td>62451</td>\n",
       "      <td>49115</td>\n",
       "      <td>53829</td>\n",
       "      <td>41438</td>\n",
       "      <td>70051</td>\n",
       "      <td>74627</td>\n",
       "      <td>66590</td>\n",
       "      <td>61814</td>\n",
       "      <td>51561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18116</th>\n",
       "      <td>ITGAM</td>\n",
       "      <td>ITGAM</td>\n",
       "      <td>6525</td>\n",
       "      <td>37556</td>\n",
       "      <td>31822</td>\n",
       "      <td>6282</td>\n",
       "      <td>6487</td>\n",
       "      <td>11115</td>\n",
       "      <td>26053</td>\n",
       "      <td>16267</td>\n",
       "      <td>...</td>\n",
       "      <td>12896</td>\n",
       "      <td>5484</td>\n",
       "      <td>4798</td>\n",
       "      <td>5626</td>\n",
       "      <td>11187</td>\n",
       "      <td>6949</td>\n",
       "      <td>7982</td>\n",
       "      <td>6373</td>\n",
       "      <td>7497</td>\n",
       "      <td>10556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18117</th>\n",
       "      <td>CD27</td>\n",
       "      <td>CD27</td>\n",
       "      <td>22402</td>\n",
       "      <td>15723</td>\n",
       "      <td>19114</td>\n",
       "      <td>25501</td>\n",
       "      <td>23853</td>\n",
       "      <td>29594</td>\n",
       "      <td>23057</td>\n",
       "      <td>14154</td>\n",
       "      <td>...</td>\n",
       "      <td>21310</td>\n",
       "      <td>28071</td>\n",
       "      <td>26988</td>\n",
       "      <td>25196</td>\n",
       "      <td>22841</td>\n",
       "      <td>25997</td>\n",
       "      <td>27928</td>\n",
       "      <td>29879</td>\n",
       "      <td>28185</td>\n",
       "      <td>17410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18118</th>\n",
       "      <td>CCR7</td>\n",
       "      <td>CCR7</td>\n",
       "      <td>34870</td>\n",
       "      <td>20814</td>\n",
       "      <td>22500</td>\n",
       "      <td>30801</td>\n",
       "      <td>29853</td>\n",
       "      <td>32579</td>\n",
       "      <td>32568</td>\n",
       "      <td>13450</td>\n",
       "      <td>...</td>\n",
       "      <td>22137</td>\n",
       "      <td>31986</td>\n",
       "      <td>21805</td>\n",
       "      <td>25120</td>\n",
       "      <td>24928</td>\n",
       "      <td>34779</td>\n",
       "      <td>35177</td>\n",
       "      <td>33277</td>\n",
       "      <td>29445</td>\n",
       "      <td>24156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18119</th>\n",
       "      <td>CD274</td>\n",
       "      <td>CD274</td>\n",
       "      <td>28441</td>\n",
       "      <td>12199</td>\n",
       "      <td>19479</td>\n",
       "      <td>45090</td>\n",
       "      <td>25951</td>\n",
       "      <td>56478</td>\n",
       "      <td>34968</td>\n",
       "      <td>11197</td>\n",
       "      <td>...</td>\n",
       "      <td>28505</td>\n",
       "      <td>49515</td>\n",
       "      <td>16595</td>\n",
       "      <td>18338</td>\n",
       "      <td>48273</td>\n",
       "      <td>27523</td>\n",
       "      <td>25738</td>\n",
       "      <td>31783</td>\n",
       "      <td>26613</td>\n",
       "      <td>10949</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18120 rows × 5758 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            feature id gene name  54, 32  98, 158  31, 39  53, 139  34, 176  \\\n",
       "0      ENSG00000187634    SAMD11       0        0       0        0        0   \n",
       "1      ENSG00000188976     NOC2L       1        0       0        0        0   \n",
       "2      ENSG00000187961    KLHL17       0        0       0        0        0   \n",
       "3      ENSG00000187583   PLEKHN1       0        0       0        0        0   \n",
       "4      ENSG00000187642     PERM1       0        0       0        0        0   \n",
       "...                ...       ...     ...      ...     ...      ...      ...   \n",
       "18115              CD4       CD4   68679    33584   39114    54248    48609   \n",
       "18116            ITGAM     ITGAM    6525    37556   31822     6282     6487   \n",
       "18117             CD27      CD27   22402    15723   19114    25501    23853   \n",
       "18118             CCR7      CCR7   34870    20814   22500    30801    29853   \n",
       "18119            CD274     CD274   28441    12199   19479    45090    25951   \n",
       "\n",
       "       56, 142  89, 127  104, 118  ...  65, 153  49, 133  73, 69  84, 102  \\\n",
       "0            0        0         0  ...        0        0       0        0   \n",
       "1            0        1         0  ...        0        0       0        0   \n",
       "2            0        0         0  ...        0        0       0        0   \n",
       "3            0        0         0  ...        0        0       0        0   \n",
       "4            0        0         0  ...        0        0       0        0   \n",
       "...        ...      ...       ...  ...      ...      ...     ...      ...   \n",
       "18115    56826    61884     31830  ...    43931    62451   49115    53829   \n",
       "18116    11115    26053     16267  ...    12896     5484    4798     5626   \n",
       "18117    29594    23057     14154  ...    21310    28071   26988    25196   \n",
       "18118    32579    32568     13450  ...    22137    31986   21805    25120   \n",
       "18119    56478    34968     11197  ...    28505    49515   16595    18338   \n",
       "\n",
       "       71, 155  40, 78  24, 108  37, 127  83, 87  86, 106  \n",
       "0            0       0        0        0       0        0  \n",
       "1            0       1        1        1       0        0  \n",
       "2            0       0        0        0       0        0  \n",
       "3            0       0        0        0       0        0  \n",
       "4            0       0        0        0       0        0  \n",
       "...        ...     ...      ...      ...     ...      ...  \n",
       "18115    41438   70051    74627    66590   61814    51561  \n",
       "18116    11187    6949     7982     6373    7497    10556  \n",
       "18117    22841   25997    27928    29879   28185    17410  \n",
       "18118    24928   34779    35177    33277   29445    24156  \n",
       "18119    48273   27523    25738    31783   26613    10949  \n",
       "\n",
       "[18120 rows x 5758 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "poo = pd.DataFrame.sparse.from_spmatrix(mat_filtered)\n",
    "poo.columns = barcodes\n",
    "poo.insert(loc=0, value=feature_ids, column=\"feature id\")\n",
    "poo.insert(loc=1, value=gene_names, column=\"gene name\")\n",
    "\n",
    "poo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "66edbc5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nimport gzip\\nimport pandas as pd\\nimport scipy.io\\n\\nbarcode_path = \"bingbing/barcodes.tsv.gz\"\\nfeatures_path = \"bingbing/features.tsv.gz\"\\nmatrix_path = \"bingbing/matrix.mtx.gz\"\\n\\nwith gzip.open(barcode_path, \\'rt\\') as f:\\n    barcode_df = pd.read_csv(f, sep=\\'\\t\\', header=None)\\n    \\nwith gzip.open(features_path, \\'rt\\') as f:\\n    features_df = pd.read_csv(f, sep=\\'\\t\\', header=None)\\n\\n# Open the .gz file for reading\\nwith gzip.open(matrix_path, \\'rb\\') as f:\\n    matrix = scipy.io.mmread(f)\\n    \\nmatrix_df = pd.DataFrame.sparse.from_spmatrix(matrix)\\n'"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "import gzip\n",
    "import pandas as pd\n",
    "import scipy.io\n",
    "\n",
    "barcode_path = \"bingbing/barcodes.tsv.gz\"\n",
    "features_path = \"bingbing/features.tsv.gz\"\n",
    "matrix_path = \"bingbing/matrix.mtx.gz\"\n",
    "\n",
    "with gzip.open(barcode_path, 'rt') as f:\n",
    "    barcode_df = pd.read_csv(f, sep='\\t', header=None)\n",
    "    \n",
    "with gzip.open(features_path, 'rt') as f:\n",
    "    features_df = pd.read_csv(f, sep='\\t', header=None)\n",
    "\n",
    "# Open the .gz file for reading\n",
    "with gzip.open(matrix_path, 'rb') as f:\n",
    "    matrix = scipy.io.mmread(f)\n",
    "    \n",
    "matrix_df = pd.DataFrame.sparse.from_spmatrix(matrix)\n",
    "\"\"\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
